packages/widget-v2/src/components/AssetChainInput.tsx

```diff
@@ -48,8 +50,63 @@ export const AssetChainInput = ({
       <Row justify="space-between">
         <StyledInput
           type="text"
-          value={value}
-          onChange={(e) => onChangeValue?.(e.target.value)}
+          value={formatNumberWithCommas(value || "")}
+          placeholder="0"
+          inputMode="numeric"
+          onChange={(e) => {
+            if (!onChangeValue) return;
+            let latest = e.target.value;
+
+            if (latest.match(/^[.,]/)) latest = `0.${latest}`; // Handle first character being a period or comma
+            latest = latest.replace(/^[0]{2,}/, "0"); // Remove leading zeros
+            latest = latest.replace(/[^\d.,]/g, ""); // Remove non-numeric and non-decimal characters
+            latest = latest.replace(/[.]{2,}/g, "."); // Remove multiple decimals
+            latest = latest.replace(/[,]{2,}/g, ","); // Remove multiple commas
+            onChangeValue?.(formatNumberWithoutCommas(latest))
+          }} onKeyDown={(event) => {
+            if (!onChangeValue) return;
+
+            if (event.key === "Escape") {
+              if (
+                event.currentTarget.selectionStart ===
+                event.currentTarget.selectionEnd
+              ) {
+                event.currentTarget.select();
+              }
+              return;
+            }
+
+            if (event.key === "ArrowUp" || event.key === "ArrowDown") {
+              let value = new BigNumber(
+                formatNumberWithoutCommas(event.currentTarget.value) || "0"
+              );
+              if (event.key === "ArrowUp") {
+                event.preventDefault();
+                if (event.shiftKey) {
+                  value = value.plus(10);
+                } else if (event.altKey || event.ctrlKey || event.metaKey) {
+                  value = value.plus(0.1);
+                } else {
+                  value = value.plus(1);
+                }
+              }
+              if (event.key === "ArrowDown") {
+                event.preventDefault();
+                if (event.shiftKey) {
+                  value = value.minus(10);
+                } else if (event.altKey || event.ctrlKey || event.metaKey) {
+                  value = value.minus(0.1);
+                } else {
+                  value = value.minus(1);
+                }
+              }
+              if (value.isNegative()) {
+                value = new BigNumber(0);
+              }
+              onChangeValue(value.toString());
+            }
+          }
+          }
```

Comment:
can we refactor onKeyDown to use switch cases? something like this
```
if (!onChangeValue) return;

  let value = new BigNumber(
    formatNumberWithoutCommas(event.currentTarget.value) || "0"
  );
  switch (event.key) {
    case "Escape":
      if (
        event.currentTarget.selectionStart ===
        event.currentTarget.selectionEnd
      ) {
        event.currentTarget.select();
      }
      return;

    case "ArrowUp":
      event.preventDefault();

      if (event.shiftKey) {
        value = value.plus(10);
      } else if (event.altKey || event.ctrlKey || event.metaKey) {
        value = value.plus(0.1);
      } else {
        value = value.plus(1);
      }
    case "ArrowDown":
      event.preventDefault();

      if (event.shiftKey) {
        value = value.minus(10);
      } else if (event.altKey || event.ctrlKey || event.metaKey) {
        value = value.minus(0.1);
      } else {
        value = value.minus(1);
      }
      
      break;

    default:
      break;
  }
  if (value.isNegative()) {
    value = new BigNumber(0);
  }
  onChangeValue(value.toString());
```

Reply:
we could also combine ArrowUp and ArrowDown code and just determine which function / plus or minus based on the direction if we wanted to

Reply:
sure

Reply:
```
    case "ArrowUp":
    case "ArrowDown":
      event.preventDefault();
      const incrementOrDecrementValue = event.key === 'ArrowUp' ? value.plus : value.minus;

      if (event.shiftKey) {
        value = incrementOrDecrementValue(10);
      } else if (event.altKey || event.ctrlKey || event.metaKey) {
        value = incrementOrDecrementValue(0.1);
      } else {
        value = incrementOrDecrementValue(1);
      }

      break;
```